### PR TITLE
[Feature] GPSR with repeated parameters

### LINKED ISSUE #217: [Bug] PSR grad wrong for repeated params
```
import torch
from pyqtorch import Z, RX, expectation
from pyqtorch import Sequence, QuantumCircuit, Observable
from pyqtorch import random_state, DiffMode

torch.manual_seed(0)

n_qubits = 2

ops = Sequence([RX(i, "theta") for i in range(n_qubits)])

obs = Observable([Z(i) for i in range(n_qubits)])

values_ad = {"theta": torch.zeros(1, requires_grad = True)}
values_psr = {"theta": torch.zeros(1, requires_grad = True)}
values_adj = {"theta": torch.zeros(1, requires_grad = True)}

circuit = QuantumCircuit(n_qubits, ops)

init_state = random_state(n_qubits)

exp_ad = expectation(circuit, state = init_state, values = values_ad, observable = obs, diff_mode = DiffMode.AD)
exp_psr = expectation(circuit, state = init_state, values = values_psr, observable = obs, diff_mode = DiffMode.GPSR)
exp_adj = expectation(circuit, state = init_state, values = values_adj, observable = obs, diff_mode = DiffMode.ADJOINT)

print(exp_ad)
print(exp_psr)
print(exp_adj)
```

```
tensor([0.1653], dtype=torch.float64, grad_fn=<SelectBackward0>)
tensor([0.1653], dtype=torch.float64, grad_fn=<PSRExpectationBackward>)
tensor([0.1653], dtype=torch.float64, grad_fn=<AdjointExpectationBackward>)
```

```
grad_ad = torch.autograd.grad(exp_ad, tuple(values_ad.values()), torch.ones_like(exp_ad))
grad_psr = torch.autograd.grad(exp_psr, tuple(values_psr.values()), torch.ones_like(exp_psr))
grad_adj = torch.autograd.grad(exp_adj, tuple(values_adj.values()), torch.ones_like(exp_adj))

print(grad_ad)
print(grad_psr)
print(grad_adj)
```

```
(tensor([-0.2279]),)
(tensor([-0.0626]),)
(tensor([-0.2279]),)
```

### PR DESCRIPTION
**Pull request by chMoussa**
Solves #217 by introducing a uuid string in parametric and HamiltonianEvolution, tweaking GPSR accordingly to use the uuid to differentiate the gate for applying the shift-rule and summing up gradient contributions. Also testing.

### CODE REVIEW

**Comment by RolandMacDoland** [comment]
Fun, it is me or this was the initial case when Dominik worked on it @jpmoutinho ? Whatever he came up with later was to avoid this uuid handling anyways.

File: pyqtorch/differentiation/gpsr.py [hunk 1]
```diff
@@ -273,18 +271,31 @@ def vjp(
...
 
-        def update_gradient(param_name: str, spectral_gap: Tensor, shift_prefac: float):
+        def update_gradient(
+            param_name: str, param_uuid: str, spectral_gap: Tensor, shift_prefac: float
+        ):
             """Update gradient of a parameter using PSR.
 
             Args:
                 param_name (str): Parameter name to compute gradient over.
+                param_uuid (str): Uuid of Parameter to hrlp compute gradient over.
```
Comment (RolandMacDoland) [inline pyqtorch/differentiation/gpsr.py:285]:
```suggestion
                param_uuid (str): Uuid of Parameter to help identify the computed gradient.
```

File: pyqtorch/primitives/parametric.py [hunk 2]
```diff
@@ -76,7 +78,9 @@ def parse_values(
             values = values or dict()
             if embedding is not None:
                 values = embedding(values)
-
+            # note: trick for using GPSR with a same param_name used in many operations
```
Comment (RolandMacDoland) [inline pyqtorch/primitives/parametric.py:81]:
```suggestion
            # note: GPSR trick when the same param_name is used in many operations
```

File: pyqtorch/quantum_operation.py [hunk 3]
```diff
@@ -213,6 +213,7 @@ def eigenvalues(
         self,
         values: dict[str, Tensor] | Tensor | None = None,
         embedding: Embedding | None = None,
+        diagonal: bool = False,
```
Comment (RolandMacDoland) [inline pyqtorch/quantum_operation.py:216]: hm it doesn't feel like these changes belong to this PR ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. Maybe the random name can be improved using a strategy ? Just a thought, please tell me what you think.

File: pyqtorch/quantum_operation.py (hunk 3 above)
Comment (chMoussa) [reply pyqtorch/quantum_operation.py:216]: No it is something I forgot in the last MR #175

File: tests/differentiation/test_gpsr.py [hunk 4]
```diff
@@ -26,13 +26,13 @@ def Hamiltonian_general(n_qubits: int = 2, batch_size: int = 1) -> torch.Tensor:
...
 
-def circuit_psr(n_qubits: int) -> QuantumCircuit:
+def circuit_psr(n_qubits: int, different_names: bool = True) -> QuantumCircuit:
     """Helper function to make an example circuit using single gap PSR."""
 
     ops = [
         pyq.RX(0, "x"),
         pyq.RY(1, "y"),
-        pyq.RX(0, "theta"),
+        pyq.RX(0, "theta" if different_names else "x"),
```
[the remark this answers is not in the document]
Comment (chMoussa) [reply tests/differentiation/test_gpsr.py:35]: Oh nothing particular here but let me see if I can cook a strategies quickly.

**Comment by jpmoutinho** [comment]
I don't recall what was the previous case. At some point this was left "not working" in PyQ on purpose because Qadence handled it through uuids.

File: pyqtorch/hamiltonians/evolution.py [hunk 5]
```diff
@@ -550,8 +556,12 @@ def jacobian(
             values = embedding(values)
 
         hamiltonian: torch.Tensor = self.create_hamiltonian(values, embedding)  # type: ignore [call-arg]
+        # note: trick for using GPSR with a same param_name used in many operations
```
Comment (RolandMacDoland) [inline pyqtorch/hamiltonians/evolution.py:559]: Probably best to harmonise with the other comment. Sorry for not having picked that up earlier.

**Review by RolandMacDoland — APPROVED** [review]
Thanks @chMoussa. Minor comment here but looking good.
